pubtools/_pulp/tasks/delete.py

```diff
@@ -0,0 +1,709 @@
+import logging
+import re
+import sys
+import attr
+
+from functools import partial
+
+from more_executors.futures import (
+    f_map,
+    f_flat_map,
+    f_zip,
+    f_return,
+    f_sequence,
+)
+from pushsource import Source, PushItem, FilePushItem, ModuleMdPushItem, RpmPushItem
+from pubtools.pulplib import (
+    Criteria,
+    Matcher,
+    RpmUnit,
+    ModulemdUnit,
+    FileUnit,
+    ErratumUnit,
+)
+
+from pubtools._pulp.services.pulp import PulpClientService
+from pubtools._pulp.services.collector import CollectorService
+from pubtools._pulp.task import PulpTask
+from pubtools._pulp.tasks.common import Publisher
+from pubtools._pulp.arguments import SplitAndExtend
+
+LOG = logging.getLogger("pubtools.pulp")
+
+step = PulpTask.step
+
+MODULEMD_REGEX = re.compile(r"^[-.+\w]+:[-.+\w]+:\d+(:[-.+\w]+){0,2}$")
+
+
+@attr.s
+class ClearedRepo(object):
+    """Represents a single repo where contents were removed."""
+
+    tasks = attr.ib()
+    """The completed Pulp tasks for removing content from this repo."""
+
+    repo = attr.ib()
+    """The repo which where content was cleared."""
+
+    content = attr.ib()
+    """The content that was cleared from the repo"""
+
+
+@attr.s
+class RemoveUnitItem(object):
+    """Represents a unit with destination repos"""
+
+    unit = attr.ib()
+    """A Pulp unit"""
+
+    repos = attr.ib()
+    """The list of repos from where the unit to be removed"""
+
+
+class Delete(PulpClientService, CollectorService, Publisher, PulpTask):
```

Comment:
Some of the methods in here look very similar or maybe copy-pasted from ClearRepo class.  Would it be possible to cut down on the code duplication?

Reply:
Yes, will move the code to a common place to be used by both the commands.